File: 207-course-schedule/solution.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
  const edges: number[][] = [];
  prerequisites.forEach(([a, b]) => {
    if (!edges[b]) edges[b] = [];
    edges[b].push(a);
  });
  const seen = new Set();
  for (let i = 0; i < numCourses; i++) {
    if (seen.has(i)) continue;
    seen.add(i);
    const stack = [{ node: i, visited: new Set([i]) }];
    while (!!stack[0]) {
      const { node, visited } = stack.pop() as {
        node: number;
        visited: Set<number>;
      };
      const courses = edges[node] || [];
      for (let c of courses) {
        if (visited.has(c)) return false;
        if (!seen.has(c)) {
          seen.add(c);
          stack.push({ node: c, visited: new Set([...visited, c]) });
        }
      }
    }
  }
  return true;
}

File: 207-course-schedule/solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> e(numCourses);
        vector<int> indeg(numCourses, 0);
        for(auto& x : prerequisites) {
            e[x[0]].push_back(x[1]);
            indeg[x[1]]++;
        }
        queue<int> qu;
        for(int i = 0; i < numCourses; i++) {
            if(indeg[i] == 0) qu.push(i);
        }
        vector<int> order;
        while(!qu.empty()) {
            auto u = qu.front();
            qu.pop();
            order.push_back(u);
            for(auto& v : e[u]) {
                if(--indeg[v] == 0) qu.push(v);
            }
        }
        return order.size() == numCourses;
    }
};
